lib/std/mem.zig

```diff
@@ -1829,6 +1829,49 @@ test "replace" {
     testing.expect(eql(u8, output[0..], "Favor reading  over writing ."));
 }
 
+/// Replace all occurences of `needle` with `replacement`.
+pub fn replaceElem(comptime T: type, slice: []T, needle: T, replacement: T) void {
+    for (slice) |e, i| {
+        if (e == needle) {
+            slice[i] = replacement;
+        }
+    }
+}
+
+/// Collapse consecutive duplicate elements into one entry.
+pub fn collapseRepeats(comptime T: type, slice: []T, needle: T) usize {
```

Comment:
Not much of a needle, it's more a `elem`ent no?
And given it squishes `slice` contents it makes sense for this to return a slice of it.

Reply:
I've changed `needle` to `elem`.

As for returning a slice, such an interface could be supported with a wrapper function at no runtime cost. In the current uses, only the resulting size is used and this same interface is propagated in a handful of places which makes the code simpler.

Reply:
You can get the old behaviour by simply adding a `.len` at the end.
Returning a slice makes the API much easier to use as `f(collapseRepeats(..))`, throwing indices around is error-prone.

Reply:
> Returning a slice makes the API much easier to use

In some cases yes.  If such a use case comes up we can add a wrapper function that does this.

Reply:
It should be the other way around, returning a slice returns a pointer _and_ the adjusted length.

Reply:
> It should be the other way around, returning a slice returns a pointer and the adjusted length.

My philosphy is to do less, and then only do more if you need to.  Going the other way around means always doing more even if you don't need to, and then undoing what you don't need.

Reply:
Another example of an API that uses this methodology is the Allocator `resizeFn` function.

Reply:
If you're adding new public-facing APIs you can't just think in terms of "do less" or think of your single use-case.
Returning the length makes using this function with sliced inputs hard to use, consider the following example:
```zig
const x = collapseRepeats(foo[bar..Baz], ...)
```
If you return a sub-slice the user is able to quickly use the modified chunk of input, while if the length is returned instead another slicing operation is required.

But given that opinions are like assholes I'll rest my case, I'm getting too old for this shit.

Reply:
Ok the fact that this is a public facing function is a good point.  How about creating a `collapseRepeatsLen` and `collapseRepeats` that wraps it?